### PR TITLE
Add new trait ReloadArmamentsBar which renders a bar to show the current reload status

### PR DESCRIPTION
**Pull request by jrb0001**

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#endregion
+
+using System;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits.Render
+{
+	[Desc("Visualizes the remaining time for a condition.")]
```
Comment (MustaphaTR) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: Copy-paste error, i guess.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 2]
```diff
@@ -811,11 +811,13 @@
     <Compile Include="Traits\TunnelEntrance.cs" />
     <Compile Include="FileSystem\InstallShieldPackage.cs" />
     <Compile Include="FileFormats\Blast.cs" />
+    <Compile Include="Traits\Render\ReloadArmamentsBar.cs" />
     <Compile Include="Orders\ForceModifiersOrderGenerator.cs" />
     <Compile Include="Widgets\Logic\Ingame\CommandBarLogic.cs" />
     <Compile Include="Widgets\Logic\Ingame\StanceSelectorLogic.cs" />
     <Compile Include="Widgets\Logic\ButtonTooltipWithDescHighlightLogic.cs" />
     <Compile Include="Traits\AutoTargetPriority.cs" />
+
```
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: nit: remove this stray newline.

File: OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		{
+			if (!self.Owner.IsAlliedWith(self.World.RenderPlayer))
+				return 0;
+
+			int count = 0;
+			float sum = 0;
+
+			foreach (var armament in armaments.Value)
+			{
+				sum += 1 - (armament.FireDelay / (float)armament.Weapon.ReloadDelay);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: Units typically use their primary and secondary weapons vs different target types, so this means that the bar will typically never go lower than half full.  Math.Min(<armament progress>) would be more intuitive.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 2 above)
Comment (jrb0001) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Fixed by rebasing.

File: OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs (hunk 3 above)
Comment (jrb0001) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: Done.

**Comment by GraionDilach** [comment]
The rebase went wrong.

**Comment by jrb0001** [comment]
It should be fixed, I guess.

File: OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+			armaments = Exts.Lazy(() => self.TraitsImplementing<Armament>()
+				.Where(a => info.Armaments.Contains(a.Info.Name)).ToArray());
+		}
+
+		float ISelectionBar.GetValue()
+		{
+			if (!self.Owner.IsAlliedWith(self.World.RenderPlayer))
+				return 0;
+
+			float min = 0;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: You should be able to simplify this right down to
```
return armaments.Value.Min(a => a.FireDelay / (float)a.Weapon.ReloadDelay);
```

File: OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		readonly ReloadArmamentsBarInfo info;
+		readonly Actor self;
+		readonly Lazy<Armament[]> armaments;
+
+		public ReloadArmamentsBar(Actor self, ReloadArmamentsBarInfo info)
+		{
+			this.self = self;
+			this.info = info;
+
+			armaments = Exts.Lazy(() => self.TraitsImplementing<Armament>()
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: This follows our (now very) old conventions for deferring trait lookups until after actor creation.
The modern way to do this is to do the lookup from `INotifyCreated.Created`, which is called after all the traits have been constructed.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: You should also account for disabled Armaments, so i'd expect that code to look something like:
```csharp

IEnumerable<Armament> activeArmaments;

void INotifyCreated.Created(Actor self)
{
	armaments = self.TraitsImplementing<Armament>().ToArray().Where(Exts.IsTraitEnabled);
}
```

**Review by pchote — COMMENTED** [review]
I have a couple more small requests, then I think this should be good to merge.

File: OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs (hunk 5 above)
Comment (jrb0001) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: But then the disabled status is only evaluated once, right? So the disabled status would have to be checked in `.GetValue()`?

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: The `Exts.IsTraitEnabled` will be evaluated every time the enumerable is enumerated.

**Comment by jrb0001** [comment]
Rebased and squashed.

File: OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+		public ReloadArmamentsBar(Actor self, ReloadArmamentsBarInfo info)
+		{
+			this.self = self;
+			this.info = info;
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			// Name check can be cashed but enabled check can't.
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: `cashed` -> `cached`

Comment (jrb0001) [inline OpenRA.Mods.Common/Traits/Render/ReloadArmamentsBar.cs]: done.

**Review by abcdefg30 — APPROVED** [review]

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/120978f37f856f4ac9b52857b00e7aecbb9f4c7c

**Pull request merged**